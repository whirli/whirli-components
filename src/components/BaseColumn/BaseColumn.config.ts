// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseColumn/BaseColumn.config';

// Package config
import { PropKeys } from './BaseColumn.constants';
import type { PropSpans, PropAlign, PropJustify, PropDirection } from './BaseColumn.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.SPANS]: {
    generateBreakpoints: true,
    classes: {
      '1': 'span-1',
      '2': 'span-2',
      '3': 'span-3',
      '4': 'span-4',
      '5': 'span-5',
      '6': 'span-6',
      '7': 'span-7',
      '8': 'span-8',
      '9': 'span-9',
      '10': 'span-10',
      '11': 'span-11',
      '12': 'span-12',
      ...config.styles[PropKeys.SPANS]?.classes,
    } as Record<PropSpans, string>,
  },
  [PropKeys.ALIGN]: {
    generateBreakpoints: true,
    classes: {
      default: 'align-default',
      center: 'align-center',
      end: 'align-end',
      between: 'align-between',
      ...config.styles[PropKeys.ALIGN]?.classes,
    } as Record<PropAlign, string>,
  },
  [PropKeys.JUSTIFY]: {
    generateBreakpoints: true,
    classes: {
      default: 'justify-default',
      center: 'justify-center',
      end: 'justify-end',
      between: 'justify-between',
      ...config.styles[PropKeys.JUSTIFY]?.classes,
    } as Record<PropJustify, string>,
  },
  [PropKeys.DIRECTION]: {
    generateBreakpoints: true,
    classes: {
      default: 'direction-default',
      row: 'direction-row',
      'row-reverse': 'direction-row-reverse',
      ...config.styles[PropKeys.DIRECTION]?.classes,
    } as Record<PropDirection, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.SPANS]: {
    required: false,
    type: [String, Object] as PropType<PropSpans | Record<string, PropSpans>>,
    default: '12',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.SPANS),
  },
  [PropKeys.ALIGN]: {
    required: false,
    type: [String, Object] as PropType<PropAlign | Record<string, PropAlign>>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.ALIGN),
  },
  [PropKeys.JUSTIFY]: {
    required: false,
    type: [String, Object] as PropType<PropJustify | Record<string, PropJustify>>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.JUSTIFY),
  },
  [PropKeys.DIRECTION]: {
    required: false,
    type: [String, Object] as PropType<PropDirection | Record<string, PropDirection>>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.DIRECTION),
  },
  ...config.props,
};
