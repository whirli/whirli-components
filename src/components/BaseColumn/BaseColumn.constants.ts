export const PropKeys: Record<string, string> = {
  SPANS: 'spans',
  ALIGN: 'align',
  JUSTIFY: 'justify',
  DIRECTION: 'direction',
  START: 'start',
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.SPANS]: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
  [PropKeys.ALIGN]: ['default', 'end', 'between'],
  [PropKeys.JUSTIFY]: ['default', 'end', 'between'],
  [PropKeys.DIRECTION]: ['default', 'row', 'row-reverse', 'column-reverse'],
  [PropKeys.START]: ['default', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
};
