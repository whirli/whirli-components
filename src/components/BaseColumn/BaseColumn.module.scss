@import '../../styles/Breakpoints.module.scss';

// default styles
.column {
  display: flex;
  flex-direction: column;
  overflow: auto;
  grid-column: span 12 / span 12;
}

.direction-default {
  flex-direction: column;
}

.direction-row {
  flex-direction: row;
}

// breakpoint styles
$packageBreakpointStyles: (
    span-1: ( grid-column: span 1 / span 1 ),
    span-2: ( grid-column: span 2 / span 2 ),
    span-3: ( grid-column: span 3 / span 3 ),
    span-4: ( grid-column: span 4 / span 4 ),
    span-5: ( grid-column: span 5 / span 5 ),
    span-6: ( grid-column: span 6 / span 6 ),
    span-7: ( grid-column: span 7 / span 7 ),
    span-8: ( grid-column: span 8 / span 8 ),
    span-9: ( grid-column: span 9 / span 9 ),
    span-10: ( grid-column: span 10 / span 10 ),
    span-11: ( grid-column: span 11 / span 11 ),
    span-12: ( grid-column: span 12 / span 12 ),

    align-default: (
      align-items: flex-start,
    ),
    align-end: (
      align-items: flex-end,
    ),
    align-center: (
      align-items: center,
    ),
    align-between: (
      align-items: space-between,
    ),

    justify-default: (
      justify-content: flex-start,
    ),
    justify-center: (
      justify-content: center,
    ),
    justify-end: (
      justify-content: flex-end,
    ),
    justify-between: (
      justify-content: space-between,
    ),
);