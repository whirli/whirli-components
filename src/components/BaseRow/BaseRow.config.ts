// @ts-ignore
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli-local/components/BaseRow/BaseRow.config';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  ...config.props,
};
