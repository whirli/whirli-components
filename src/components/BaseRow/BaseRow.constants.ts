export const PropKeys: Record<string, string> = {
  DIRECTION: 'direction',
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.DIRECTION]: ['row', 'row-reverse'],
};
