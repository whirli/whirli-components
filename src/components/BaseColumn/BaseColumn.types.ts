import { PropKeys, PropValues } from './BaseColumn.constants';

const propSpans = PropValues[PropKeys.SPANS];
const propAlign = PropValues[PropKeys.ALIGN];
const propJustify = PropValues[PropKeys.JUSTIFY];
const propDirection = PropValues[PropKeys.DIRECTION];

export type PropSpans = typeof propSpans[number];
export type PropAlign = typeof propAlign[number];
export type PropJustify = typeof propJustify[number];
export type PropDirection = typeof propDirection[number];
