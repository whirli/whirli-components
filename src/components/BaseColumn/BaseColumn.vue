<template>
  <div :class="classes"><slot /></div>
</template>

<script setup lang="ts">
// Vue
// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseColumn/BaseColumn.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseColumn.config';

const ComponentStyles = ConfigStyles;

const props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const classes = [styles.column, ...makeClasses(ComponentStyles, props, styles)];
</script>
