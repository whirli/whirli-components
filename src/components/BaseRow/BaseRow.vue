<template>
  <BaseColumn direction="row" v-bind="$attrs" :class="classes"><slot /></BaseColumn>
</template>

<script setup="props" lang="ts">
// Vue
import { defineProps } from 'vue';

// Styles
import styles from '@whirli-local/components/BaseRow/BaseRow.module.scss';

// Data
import { ConfigStyles, ConfigProps } from './BaseRow.config';

// Components
import BaseColumn from '../BaseColumn/BaseColumn.vue';

// Types
import { ComponentStyles as ComponentStylesInterface } from '../../@types/components';
import { Props } from '../../@types/props';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

const props: Props = defineProps(ConfigProps);

// Classes
import useClasses from '../../@use/class';
const { makeClasses } = useClasses();
const classes = makeClasses(ComponentStyles, props, styles);
</script>
