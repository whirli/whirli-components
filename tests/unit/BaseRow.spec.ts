import { mount } from '@vue/test-utils';

import BaseRow from '@/components/BaseRow/BaseRow.vue';
import BaseColumn from '@/components/BaseColumn/BaseColumn.vue';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let wrapper: any;
const text = 'Hello World';

describe('BaseRow.vue', () => {
  beforeEach(() => {
    wrapper = mount(BaseRow, {
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a column component', () => {
    expect(wrapper.findComponent(BaseColumn).exists()).toBeTruthy();
  });

  it('displays the correct text', () => {
    expect(wrapper.findComponent(BaseColumn).text()).toMatch(text);
  });
});
