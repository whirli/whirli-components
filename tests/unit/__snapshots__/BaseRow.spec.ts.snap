// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseRow.vue renders the component 1`] = `
<div
  class=""
>
  
  
  Hello World
  
  
</div>
`;
